google-cloud-datastore/src/test/java/com/google/cloud/datastore/it/ITDatastoreTest.java

```diff
@@ -0,0 +1,670 @@
+  private static final FullEntity<IncompleteKey> PARTIAL_ENTITY2 =
+      FullEntity.builder(PARTIAL_ENTITY1).remove("str").set("bool", true)
+          .set("list", LIST_VALUE1.get()).build();
+  private static final FullEntity<IncompleteKey> PARTIAL_ENTITY3 =
+      FullEntity.builder(PARTIAL_ENTITY1).key(IncompleteKey.builder(PROJECT_ID, KIND3).build())
+          .build();
+  private static final Entity ENTITY1 = Entity.builder(KEY1)
+      .set("str", STR_VALUE)
+      .set("date", DATE_TIME_VALUE)
+      .set("latLng", LAT_LNG_VALUE)
+      .set("bool", BOOL_VALUE)
+      .set("partial1", EntityValue.of(PARTIAL_ENTITY1))
+      .set("list", LIST_VALUE2)
+      .set("emptyList", EMPTY_LIST_VALUE)
+      .build();
+  private static final Entity ENTITY2 = Entity.builder(ENTITY1).key(KEY2).remove("str")
+      .set("name", "Dan").setNull("null").set("age", 20).build();
+  private static final Entity ENTITY3 = Entity.builder(ENTITY1).key(KEY3).remove("str")
+      .set("null", NULL_VALUE).set("partial1", PARTIAL_ENTITY2).set("partial2", ENTITY2).build();
+
+  @AfterClass
+  public static void afterClass() {
+    HELPER.deleteNamespace();
+  }
+
+  @Before
+  public void setUp() {
+    DATASTORE.put(ENTITY1, ENTITY2);
+  }
+
+  @After
+  public void tearDown() {
+    DATASTORE.delete(KEY1, KEY2, KEY3);
+  }
+
+  @Test
+  public void testNewTransactionCommit() {
+    Transaction transaction = DATASTORE.newTransaction();
+    transaction.add(ENTITY3);
+    Entity entity2 = Entity.builder(ENTITY2)
+        .clear()
+        .setNull("bla")
+        .build();
+    transaction.update(entity2);
+    transaction.delete(KEY1);
+    transaction.commit();
+
+    List<Entity> list = DATASTORE.fetch(KEY1, KEY2, KEY3);
+    assertNull(list.get(0));
+    assertEquals(entity2, list.get(1));
+    assertEquals(ENTITY3, list.get(2));
+    assertEquals(3, list.size());
+
+    try {
+      transaction.commit();
+      fail("Expecting a failure");
+    } catch (DatastoreException ex) {
+      // expected to fail
+    }
+
+    try {
+      transaction.rollback();
+      fail("Expecting a failure");
+    } catch (DatastoreException ex) {
+      // expected to fail
+    }
+  }
+
+  @Test
+  public void testTransactionWithRead() {
+    Transaction transaction = DATASTORE.newTransaction();
+    assertNull(transaction.get(KEY3));
+    transaction.add(ENTITY3);
+    transaction.commit();
+    assertEquals(ENTITY3, DATASTORE.get(KEY3));
+
+    transaction = DATASTORE.newTransaction();
+    assertEquals(ENTITY3, transaction.get(KEY3));
+    // update entity3 during the transaction
+    DATASTORE.put(Entity.builder(ENTITY2).clear().set("from", "datastore").build());
+    transaction.update(Entity.builder(ENTITY2).clear().set("from", "transaction").build());
+    try {
+      transaction.commit();
+      fail("Expecting a failure");
+    } catch (DatastoreException expected) {
+      assertEquals("ABORTED", expected.reason());
+    }
+  }
+
+  @Test
```

Comment:
Should there be an overall time limit for this to complete?  (I seem to recall a very long period of time, but it doesn't appear to be published anywhere anymore).